app/retail/views.py

```diff
@@ -742,84 +744,62 @@ def results(request, keyword=None):
     elif 'grant:' in what:
-        activity_pks = ActivityIndex.objects.filter(key=what)
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key=what)
 
     # all Grants
     if what == 'all_grants':
-        activity_pks = ActivityIndex.objects.filter(key__startswith='grant:')
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key__startswith='grant:')
 
     # kudos
     if (
         what in ['kudos']
     ):
-        activity_pks = ActivityIndex.objects.filter(key__startswith='kudo:')
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key__startswith='kudo:')
     elif 'kudos:' in what:
-        activity_pks = ActivityIndex.objects.filter(key=what.replace('kudos', 'kudo'))
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key=what.replace('kudos', 'kudo'))
 
     # hackathon project
     if 'project:' in what:
-        activity_pks = ActivityIndex.objects.filter(key=what)
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key=what)
 
     # tribes
     if 'tribe:' in what:
         handle = what[6:]
         profile = Profile.objects.filter(handle=handle).first()
-        filter_applied = True
         if profile:
-            activity_pks = ActivityIndex.objects.filter(key=f'profile:{profile.pk}')
+            activities = Activity.objects.filter(activities_index__key=f'profile:{profile.pk}')
 
     # hackathon activity
     if 'hackathon:' in what:
-        activity_pks = ActivityIndex.objects.filter(key=what)
-        filter_applied = True
+        activities = Activity.objects.filter(activities_index__key=what)
 
     # single activity
     if 'activity:' in what:
-        activity_pks = [what.replace('activity:', '')]
-        filter_applied = True
+        activities = [what.replace('activity:', '')]
 
     # Defaults
-    if not activity_pks and not filter_applied:
-        activity_pks = ActivityIndex.objects.all()
-
-    # Are the pks already in a list?
-    if not isinstance(activity_pks, list):
-        # Order the activity index data
-        activity_pks = activity_pks.order_by('-id')
-        # Pagination is done here
-        if page:
-            start_index = (page-1) * page_size
-            end_index = page * page_size
-            activity_pks = activity_pks[start_index:end_index].values_list('activity_id', flat=True)
-        else:
-            activity_pks = activity_pks.values_list('activity_id', flat=True)
+    if not activities:
+        activities = Activity.objects.exclude(activities_index__key__isnull=True)
 
     # Cross-ref the activity_pks->activity_id with the Activity objects
-    activities = Activity.objects.filter(pk__in=list(activity_pks),hidden=False).order_by('-created_on')
+    activities = activities.filter(hidden=False).order_by('-created_on')[start_index:end_index]
```

Comment:
`activities` is a list on ln 795 - we would want to wrap this to make sure `activities` can be filtered before applying this `filter`

Reply:
nice catch !